packages/@aws-cdk/aws-eks/lib/cluster-resource.ts

```diff
@@ -32,51 +30,66 @@ export class ClusterResource extends Construct {
-      code: lambda.Code.fromAsset(path.join(__dirname, 'cluster-resource')),
-      runtime: lambda.Runtime.PYTHON_3_7,
-      handler: 'index.handler',
-      timeout: Duration.minutes(15),
-      memorySize: 512,
-      layers: [ KubectlLayer.getOrCreate(this) ],
-    });
+    const provider = ClusterResourceProvider.getOrCreate(this);
 
     if (!props.roleArn) {
       throw new Error(`"roleArn" is required`);
     }
 
-    // since we don't know the cluster name at this point, we must give this role star resource permissions
-    handler.addToRolePolicy(new iam.PolicyStatement({
-      actions: [ 'eks:CreateCluster', 'eks:DescribeCluster', 'eks:DeleteCluster', 'eks:UpdateClusterVersion' ],
-      resources: [ '*' ]
-    }));
+    // the role used to create the cluster. this becomes the administrator role
+    // of the cluster.
+    this.creationRole = new iam.Role(this, 'CreationRole', {
+      assumedBy: new iam.CompositePrincipal(...provider.roles.map(x => new iam.ArnPrincipal(x.roleArn)))
+    });
 
     // the CreateCluster API will allow the cluster to assume this role, so we
     // need to allow the lambda execution role to pass it.
-    handler.addToRolePolicy(new iam.PolicyStatement({
+    this.creationRole.addToPolicy(new iam.PolicyStatement({
       actions: [ 'iam:PassRole' ],
       resources: [ props.roleArn ]
     }));
 
+    // since we don't know the cluster name at this point, we must give this role star resource permissions
+    this.creationRole.addToPolicy(new iam.PolicyStatement({
+      actions: [ 'eks:CreateCluster', 'eks:DescribeCluster', 'eks:DeleteCluster', 'eks:UpdateClusterVersion', 'eks:UpdateClusterConfig' ],
+      resources: [ '*' ]
+    }));
+
     const resource = new cfn.CustomResource(this, 'Resource', {
       resourceType: ClusterResource.RESOURCE_TYPE,
-      provider: cfn.CustomResourceProvider.lambda(handler),
+      provider: provider.provider,
       properties: {
-        Config: props
+        Config: props,
+        AssumeRoleArn: this.creationRole.roleArn
       }
     });
 
+    resource.node.addDependency(this.creationRole);
+
     this.ref = resource.ref;
     this.attrEndpoint = Token.asString(resource.getAtt('Endpoint'));
     this.attrArn = Token.asString(resource.getAtt('Arn'));
     this.attrCertificateAuthorityData = Token.asString(resource.getAtt('CertificateAuthorityData'));
-    this.creationRole = handler.role!;
+  }
+
+  /**
+   * Returns the ARN of the cluster creation role and grants `trustedRole`
+   * permissions to assume this role.
+   */
+  public getCreationRoleArn(trustedRole: iam.IRole): string {
+    if (!this.trustedPrincipals.includes(trustedRole.roleArn)) {
+      this.creationRole.assumeRolePolicy?.addStatements(new iam.PolicyStatement({
+        actions: [ 'sts:AssumeRole' ],
+        principals: [ new iam.ArnPrincipal(trustedRole.roleArn) ]
+      }));
+      this.trustedPrincipals.push(trustedRole.roleArn);
```

Comment:
This is a lie if `this.creationRole.assumeRolePolicy` is undefined, as the above statement is defensive against...

Reply:
fair enough, added a defensive check